gui/storage/forms.py

```diff
@@ -1635,54 +1695,85 @@ def __init__(self, *args, **kwargs):
             with client as c:
                 stat = c.call('filesystem.stat', path)
 
-                self.fields['mp_acl'].initial = stat['acl']
-
-                if stat['acl'] == 'windows':
+                """
+                Disable the mode editor if the path has an extended ACL.
+                """
+                if stat['acl']:
                     self.fields['mp_mode'].widget.attrs['disabled'] = 'disabled'
 
                 self.fields['mp_mode'].initial = "%.3o" % stat['mode']
                 self.fields['mp_user'].initial = stat['user']
                 self.fields['mp_group'].initial = stat['group']
-
         self.fields['mp_acl'].widget.attrs['onChange'] = "mpAclChange(this);"
 
     def commit(self, path):
-
-        with client as c:
-            dataset = c.call('pool.dataset.query', [['mountpoint', '=', path.rstrip('/')]], {'get': True})
-
         kwargs = {}
 
+        uid = -1
+        gid = -1
         if self.cleaned_data.get('mp_group_en'):
             kwargs['group'] = self.cleaned_data['mp_group']
+            with client as c:
+                gid = (c.call('dscache.get_uncached_group', kwargs['group']))['gr_gid']
 
         if self.cleaned_data.get('mp_user_en'):
             kwargs['user'] = self.cleaned_data['mp_user']
+            with client as c:
+                uid = (c.call('dscache.get_uncached_user', kwargs['user']))['pw_uid']
 
-        kwargs['acl'] = self.cleaned_data['mp_acl'].upper()
-        if kwargs['acl'] != 'WINDOWS' and self.cleaned_data.get('mp_mode_en'):
+        if self.cleaned_data.get('mp_mode_en'):
             kwargs['mode'] = str(self.cleaned_data['mp_mode'])
 
-        kwargs['recursive'] = self.cleaned_data['mp_recursive']
+        kwargs['traverse'] = self.cleaned_data['mp_traverse']
+        action = self.cleaned_data['mp_acl']
 
-        with client as c:
-            try:
-                c.call('pool.dataset.permission', dataset['id'], kwargs)
+        if action == 'applydefault':
+            with client as c:
+                c.call(
+                    'filesystem.setacl',
+                    path,
+                    [
+                        {
+                            "tag": "owner@",
+                            "id": None,
+                            "type": "ALLOW",
+                            "perms": {"BASIC": "FULL_CONTROL"},
+                            "flags": {"BASIC": "INHERIT"}
+                        },
+                        {
+                            "tag": "group@",
+                            "id": None,
+                            "type": "ALLOW",
+                            "perms": {"BASIC": "FULL_CONTROL"},
+                            "flags": {"BASIC": "INHERIT"}
+                        },
+                    ],
+                    uid,
+                    gid,
```

Comment:
This seems to be a rather complicated call to "reset" ACL. Are we expecting new UI to use the same?

Perhaps it could be worthwhile to create a wrapper or another mutually-exclusive attribute to perform that action?

Thoughts?

Reply:
I'll simplify by just having it reflected as the default ACL in pool.dataset.permission (which is a sort of wrapper around filesystem.setacl and filesystem.setperm).